File: src/routes/api/smc-strategy-config.ts
import { createFileRoute } from "@tanstack/react-router";
import { getCFEnv } from "@/lib/cloudflare-env";

/**
 * SMC Strategy Configuration endpoint.
 * 
 * The SMC engine has tunable parameters (ATR period, swing window, OB thresholds, etc.)
 * that control how it detects structure, order blocks, FVGs, and sweeps.
 * 
 * GET returns the current config (or defaults if none stored).
 * PATCH updates the config — used by both manual edits and Hermes upgrades.
 */

const DEFAULT_CONFIG = {
  atr_period: 14,
  swing_window: 3,
  ob_impulse_mult: 1.5,
  ob_max_age: 60,
  fvg_max_age: 60,
  sweep_max_age: 30,
  bos_min_bars: 0,
  retest_min_touches: 2,
  sl_pips: 30,
  tp1_rr: 1.5,
  tp2_rr: 2.0,
  confluence_weights: {
    ob_retest: 2.0,
    sweep: 2.0,
    choch: 1.5,
    bos: 1.0,
    bias: 1.0,
    fvg: 0.5,
    zone: 0.5,
  },
};

const PARAM_META: Record<string, { label: string; min: number; max: number; step: number; description: string }> = {
  atr_period: { label: "ATR Period", min: 5, max: 50, step: 1, description: "ATR lookback period. Higher = smoother, less sensitive to recent volatility." },
  swing_window: { label: "Swing Window", min: 2, max: 10, step: 1, description: "Fractal pivot detection window. Higher = fewer but more significant swing points." },
  ob_impulse_mult: { label: "OB Impulse Multiplier", min: 0.5, max: 4.0, step: 0.1, description: "Order block impulse threshold in ATR multiples. Higher = only the strongest order blocks." },
  ob_max_age: { label: "OB Max Age", min: 20, max: 200, step: 5, description: "Maximum bars to look back for order blocks. Higher = more historical levels." },
  fvg_max_age: { label: "FVG Max Age", min: 20, max: 200, step: 5, description: "Maximum bars to look back for fair value gaps." },
  sweep_max_age: { label: "Sweep Max Age", min: 10, max: 100, step: 5, description: "Maximum bars to look back for liquidity sweeps." },
  bos_min_bars: { label: "BOS Min Bars", min: 0, max: 10, step: 1, description: "Minimum bars before confirming a break of structure." },
  retest_min_touches: { label: "Retest Min Touches", min: 1, max: 5, step: 1, description: "Minimum touches of breakout boundary for a valid channel." },
  sl_pips: { label: "Stop Loss (pips)", min: 10, max: 100, step: 1, description: "Default stop loss distance in pips." },
  tp1_rr: { label: "TP1 R:R", min: 1.0, max: 5.0, step: 0.1, description: "Take profit 1 risk:reward multiple." },
  tp2_rr: { label: "TP2 R:R", min: 1.0, max: 5.0, step: 0.1, description: "Take profit 2 risk:reward multiple." },
};

export const Route = createFileRoute("/api/smc-strategy-config")({
  server: {
    handlers: {
      GET: async () => {
        const env = getCFEnv();
        if (!env) return Response.json({ config: DEFAULT_CONFIG, defaults: DEFAULT_CONFIG, param_meta: PARAM_META });

        try {
          // Ensure table exists
          await env.DB.prepare(`
            CREATE TABLE IF NOT EXISTS smc_strategy_config (
              key TEXT PRIMARY KEY,
              value TEXT NOT NULL,
              updated_at TEXT DEFAULT (strftime('%Y-%m-%dT%H:%M:%SZ','now'))
            )
          `).bind().run().catch(() => {});

          const { results } = await env.DB.prepare(
            "SELECT key, value FROM smc_strategy_config"
          ).bind().all();

          const stored: Record<string, unknown> = {};
          for (const row of results as any[]) {
            try {
              stored[row.key] = JSON.parse(row.value);
            } catch {
              stored[row.key] = row.value;
            }
          }

          // Merge with defaults (stored overrides default)
          const config = { ...DEFAULT_CONFIG, ...stored };
          // Deep merge confluence_weights
          if (stored["confluence_weights"]) {
            config["confluence_weights"] = { ...DEFAULT_CONFIG.confluence_weights, ...(stored["confluence_weights"] as Record<string, number>) };
          }

          return Response.json({ config, defaults: DEFAULT_CONFIG, param_meta: PARAM_META });
        } catch {
          return Response.json({ config: DEFAULT_CONFIG, defaults: DEFAULT_CONFIG, param_meta: PARAM_META });
        }
      },

      PATCH: async ({ request }) => {
        const env = getCFEnv();
        if (!env) return new Response("Service unavailable", { status: 503 });

        try {
          const body = await request.json() as Record<string, unknown>;

          // Ensure table exists
          await env.DB.prepare(`
            CREATE TABLE IF NOT EXISTS smc_strategy_config (
              key TEXT PRIMARY KEY,
              value TEXT NOT NULL,
              updated_at TEXT DEFAULT (strftime('%Y-%m-%dT%H:%M:%SZ','now'))
            )
          `).bind().run().catch(() => {});

          // Store each key-value pair
          for (const [key, value] of Object.entries(body)) {
            await env.DB.prepare(
              "INSERT OR REPLACE INTO smc_strategy_config (key, value, updated_at) VALUES (?, ?, strftime('%Y-%m-%dT%H:%M:%SZ','now'))"
            ).bind(key, JSON.stringify(value)).run();
          }

          return Response.json({ ok: true });
        } catch (e) {
          return Response.json({ ok: false, error: String(e) }, { status: 500 });
        }
      },

      // Reset to defaults
      DELETE: async () => {
        const env = getCFEnv();
        if (!env) return new Response("Service unavailable", { status: 503 });

        try {
          await env.DB.prepare("DELETE FROM smc_strategy_config").bind().run();
          return Response.json({ ok: true, config: DEFAULT_CONFIG });
        } catch (e) {
          return Response.json({ ok: false, error: String(e) }, { status: 500 });
        }
      },
    },
  },
});

File: scripts/smc_strategy_config.py
#!/usr/bin/env python3
"""
GizzyFx SMC Strategy Configuration — stored in D1 so it persists across deploys.

The strategy parameters are the knobs that control how the SMC engine detects
structure, order blocks, FVGs, and sweeps. Users can tune these manually or
ask Hermes to suggest improvements based on current market behavior.

Parameters:
  atr_period         — ATR lookback (default 14). Higher = smoother, less sensitive.
  swing_window       — Fractal pivot window (default 3). Higher = fewer, more significant swings.
  ob_impulse_mult    — Order block impulse threshold in ATR multiples (default 1.5). Higher = only strongest OB.
  ob_max_age         — Max bars to look back for OB (default 60). Higher = more historical OB.
  fvg_max_age        — Max bars to look back for FVG (default 60).
  sweep_max_age      — Max bars to look back for liquidity sweeps (default 30).
  bos_min_bars       — Min bars before confirming BOS (default 0).
  retest_min_touches — Min touches for valid channel (default 2).
  sl_pips            — Default stop loss in pips (default 30).
  tp1_rr             — TP1 risk:reward multiple (default 1.5).
  tp2_rr             — TP2 risk:reward multiple (default 2.0).
  confluence_weights — Scoring weights for each SMC factor.
"""

import os, json, time
from datetime import datetime, timezone

BASE_URL = os.environ.get("GIZZYFX_BASE_URL", "https://gizzyfxstrategy.dpdns.org")
API_KEY = os.environ.get("GIZZYFX_API_KEY", "")

DEFAULT_CONFIG = {
    "atr_period": 14,
    "swing_window": 3,
    "ob_impulse_mult": 1.5,
    "ob_max_age": 60,
    "fvg_max_age": 60,
    "sweep_max_age": 30,
    "bos_min_bars": 0,
    "retest_min_touches": 2,
    "sl_pips": 30,
    "tp1_rr": 1.5,
    "tp2_rr": 2.0,
    "confluence_weights": {
        "ob_retest": 2.0,
        "sweep": 2.0,
        "choch": 1.5,
        "bos": 1.0,
        "bias": 1.0,
        "fvg": 0.5,
        "zone": 0.5,
    },
}

PARAM_DESCRIPTIONS = {
    "atr_period": "ATR lookback period. Higher = smoother, less sensitive to recent volatility.",
    "swing_window": "Fractal pivot detection window. Higher = fewer but more significant swing points.",
    "ob_impulse_mult": "Order block impulse threshold (ATR multiples). Higher = only the strongest order blocks.",
    "ob_max_age": "Maximum bars to look back for order blocks. Higher = more historical levels.",
    "fvg_max_age": "Maximum bars to look back for fair value gaps.",
    "sweep_max_age": "Maximum bars to look back for liquidity sweeps.",
    "bos_min_bars": "Minimum bars before confirming a break of structure.",
    "retest_min_touches": "Minimum touches of breakout boundary for a valid channel.",
    "sl_pips": "Default stop loss distance in pips.",
    "tp1_rr": "Take profit 1 risk:reward multiple.",
    "tp2_rr": "Take profit 2 risk:reward multiple.",
    "confluence_weights": "Scoring weights for each SMC confluence factor.",
}


def get_config():
    """Read current config from D1."""
    try:
        r = requests.get(f"{BASE_URL}/api/smc-strategy-config", headers={"X-Hermes-Key": API_KEY}, timeout=10)
        if r.ok:
            data = r.json()
            if data.get("config"):
                return data["config"]
    except Exception:
        pass
    return DEFAULT_CONFIG


def save_config(config):
    """Write config to D1."""
    try:
        r = requests.patch(
            f"{BASE_URL}/api/smc-strategy-config",
            headers={"X-Hermes-Key": API_KEY, "Content-Type": "application/json"},
            json=config,
            timeout=10,
        )
        return r.ok
    except Exception:
        return False


if __name__ == "__main__":
    import sys
    if len(sys.argv) > 1 and sys.argv[1] == "show":
        cfg = get_config()
        print(json.dumps(cfg, indent=2))
    elif len(sys.argv) > 1 and sys.argv[1] == "reset":
        save_config(DEFAULT_CONFIG)
        print("Reset to defaults")
    else:
        print("Usage: python3 smc_strategy_config.py [show|reset]")
